{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "7af49a27-c44a-4583-81ab-0c56bc064160",
   "metadata": {},
   "source": [
    "# De Bruijn graphs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4c9944d6-194f-418a-8d64-546ceb981b02",
   "metadata": {},
   "source": [
    "You will be implementing one of the primary assembly algorithms from short-read data that is used today. We will implement a simple form of the algorithm where we _assume perfect sequencing_. That is, everything is sequenced exactly once and there are no errors or variants in the sequencing. \n",
    "\n",
    "A graph is composed of **nodes** and **edges** and we will need to develop a data strcture to track edges between nodes in our graph. We have provided the basic class structure as well as descriptions of functions to `add_edge` and `remove_edge` from the graph. You will need to implement these functions in order to then build the de Bruijn graph. \n",
    "\n",
    "In our implementation below, we use a `defaultdict` data structure to hold a list of all edges in the graph where all \"right\" nodes connected to a \"left\" node are stored in a list for that node.\n",
    "\n",
    "```\n",
    "build_debruijn_graph:\n",
    "define substring length k and input string\n",
    "For each k-length substring of input:\n",
    "  split k mer into left and right k-1 mer\n",
    "  add k-1 mers as nodes with a directed edge from left k-1 mer to right k-1 mer\n",
    "```\n",
    "\n",
    "---\n",
    "## Eulerian walk\n",
    "\n",
    "To continue our implementation from last class, we will use our De Bruijn graph to output a valid sequence from the assembly. This is implemented as a recursive algorithm by considering all valid edges. You will notice that as you change $k$, we are able to better recapitulate our sequence depending on how repetitive it is. In a more complex implementation of a Eulerian walk there are heuristics and defined rules for determining the validity of traversing a specific edge in the graph to result in a full graph-traversal. One of these methods is to traverse the graph in a depth first manner to avoid sectioning off any part of the graph in the traversal. In our implementation we will ignore these for simplicity.\n",
    "\n",
    "```\n",
    "eulerian_walk:\n",
    "Beginning at first_node as node\n",
    "\n",
    "For node:\n",
    "    follow a random valid edge from node\n",
    "    remove edge\n",
    "    recurse\n",
    "```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "70cb7948-5442-4823-9bad-7beb26ce9af3",
   "metadata": {},
   "source": [
    "---\n",
    "# Toy Example"
   ]
  },
  {
   "cell_type": "code",
   "id": "c4f1294d",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2026-02-20T01:00:08.145595Z",
     "start_time": "2026-02-20T01:00:08.103267Z"
    }
   },
   "source": [
    "import os\n",
    "print(os.getcwd())\n",
    "from utils import read_fastq, assemble_mouse_genome\n"
   ],
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/chanteralazard/Projects/Northeastern_Masters/BINF6250/EmpireOfDirt/project04\n"
     ]
    }
   ],
   "execution_count": 6
  },
  {
   "cell_type": "code",
   "id": "e1e6c2c3-3341-4c32-b813-8c602933f73b",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2026-02-21T02:17:05.304651Z",
     "start_time": "2026-02-21T02:17:05.190301Z"
    }
   },
   "source": [
    "print(\"=\"*60)\n",
    "print(\"EXAMPLE 1: Assembling 9 overlapping reads into one contig\")\n",
    "print(\"=\"*60)\n",
    "\n",
    "toy_reads_1 = [\n",
    "    \"ATGGCGTACG\",  # Read 1\n",
    "    \"GGCGTACGTT\",  # Read 2: overlaps with Read 1\n",
    "    \"CGTACGTTAC\",  # Read 3: overlaps with Read 2\n",
    "    \"TACGTTACCA\",  # Read 4: overlaps with Read 3\n",
    "    \"CGTTACCATG\",  # Read 5: overlaps with Read 4\n",
    "    \"TTACCATGGG\",  # Read 6: overlaps with Read 5\n",
    "    \"ACCATGGGCC\",  # Read 7: overlaps with Read 6\n",
    "    \"CATGGGCCTA\",  # Read 8: overlaps with Read 7\n",
    "    \"TGGGCCTAAA\"   # Read 9: overlaps with Read 8\n",
    "]\n",
    "\n",
    "print(f\"\\nInput: {len(toy_reads_1)} reads\")\n",
    "print(\"First read:  \", toy_reads_1[0])\n",
    "print(\"Last read:   \", toy_reads_1[-1])\n",
    "print(f\"\\nBuilding De Bruijn graph with k=6...\")\n",
    "\n",
    "from de_bruijn_module import DeBruijnGraph\n",
    "toy_graph = DeBruijnGraph(toy_reads_1, 6, 43375)\n",
    "toy_graph_node = toy_graph.get_starting_node()\n",
    "#print(toy_graph_node)\n",
    "euler_walk_toy = toy_graph.eulerian_walk(toy_graph_node, toy_graph.graph)\n",
    "print(f\"\\n{euler_walk_toy}\")"
   ],
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "============================================================\n",
      "EXAMPLE 1: Assembling 9 overlapping reads into one contig\n",
      "============================================================\n",
      "\n",
      "Input: 9 reads\n",
      "First read:   ATGGCGTACG\n",
      "Last read:    TGGGCCTAAA\n",
      "\n",
      "Building De Bruijn graph with k=6...\n",
      "defaultdict(<class 'list'>, {'TGGCG': ['ATGGC'], 'GGCGT': ['TGGCG'], 'GCGTA': ['GGCGT', 'GGCGT'], 'CGTAC': ['GCGTA', 'GCGTA'], 'GTACG': ['CGTAC', 'CGTAC', 'CGTAC'], 'TACGT': ['GTACG', 'GTACG'], 'ACGTT': ['TACGT', 'TACGT', 'TACGT'], 'CGTTA': ['ACGTT', 'ACGTT'], 'GTTAC': ['CGTTA', 'CGTTA', 'CGTTA'], 'TTACC': ['GTTAC', 'GTTAC'], 'TACCA': ['TTACC', 'TTACC', 'TTACC'], 'ACCAT': ['TACCA', 'TACCA'], 'CCATG': ['ACCAT', 'ACCAT', 'ACCAT'], 'CATGG': ['CCATG', 'CCATG'], 'ATGGG': ['CATGG', 'CATGG', 'CATGG'], 'TGGGC': ['ATGGG', 'ATGGG'], 'GGGCC': ['TGGGC', 'TGGGC', 'TGGGC'], 'GGCCT': ['GGGCC', 'GGGCC'], 'GCCTA': ['GGCCT', 'GGCCT'], 'CCTAA': ['GCCTA'], 'CTAAA': ['CCTAA']})\n",
      "\n",
      "[np.str_('ATGGC'), np.str_('TGGCG'), np.str_('GGCGT'), np.str_('GCGTA'), np.str_('CGTAC'), np.str_('GTACG'), np.str_('TACGT'), np.str_('ACGTT'), np.str_('CGTTA'), np.str_('GTTAC'), np.str_('TTACC'), np.str_('TACCA'), np.str_('ACCAT'), np.str_('CCATG'), np.str_('CATGG'), np.str_('ATGGG')]\n"
     ]
    }
   ],
   "execution_count": 1
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2026-02-20T22:17:57.784980Z",
     "start_time": "2026-02-20T22:17:57.740117Z"
    }
   },
   "cell_type": "code",
   "source": "print(toy_graph_node)",
   "id": "18701162ee24d06c",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GTACG\n"
     ]
    }
   ],
   "execution_count": 6
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2026-02-20T22:18:00.057499Z",
     "start_time": "2026-02-20T22:18:00.016515Z"
    }
   },
   "cell_type": "code",
   "source": [
    "print(toy_graph.tour_to_sequence(euler_walk_toy))\n",
    "#ATGGCGTACGTTACCATG\n",
    "print(f\"\\nLength of tour: {len(toy_graph.tour_to_sequence(euler_walk_toy))}\")"
   ],
   "id": "c977419c085fac8f",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ATGGCGTACG\n",
      "\n",
      "Length of tour: 10\n"
     ]
    }
   ],
   "execution_count": 7
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2026-02-21T02:17:09.625302Z",
     "start_time": "2026-02-21T02:17:09.593720Z"
    }
   },
   "cell_type": "code",
   "source": [
    "contigs = toy_graph.assemble_contigs(43375)\n",
    "print(contigs)\n"
   ],
   "id": "26710de9c30a8adf",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['GGCGTACGTTACCA', 'TTACCATGGGCCTA', 'TACGTT', 'CATGGGCC', 'ACCATG', 'CGTTAC', 'TGGGCCTAAA', 'CGTACG']\n"
     ]
    }
   ],
   "execution_count": 2
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2026-02-21T02:17:10.769811Z",
     "start_time": "2026-02-21T02:17:10.747999Z"
    }
   },
   "cell_type": "code",
   "source": [
    "toy_graph.get_assembly_stats(contigs)\n",
    "toy_graph.write_fasta(contigs, \"toy_assembly.fasta\", \"assembled_contigs\")"
   ],
   "id": "598f59704599411b",
   "outputs": [],
   "execution_count": 3
  },
  {
   "cell_type": "markdown",
   "id": "a5627a52-aa5a-49ae-8f62-38320a5f7182",
   "metadata": {},
   "source": [
    "---\n",
    "# Real data\n",
    "This section utilizes a FASTQ file with 10 million \"perfect\" 150 bp reads simulated\n",
    "from the mouse genome (`GRCm39`) and tasks your program to ingest, assemble, and traverse\n",
    "the genome with statistical output report."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7aefa0cb-5ab2-462b-ac7b-d574e03588c7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "================================================================================\n",
      "MOUSE GENOME DE BRUIJN GRAPH ASSEMBLY\n",
      "Student Assignment Driver Program\n",
      "================================================================================\n",
      "\n",
      "================================================================================\n",
      "MOUSE GENOME ASSEMBLY PIPELINE\n",
      "================================================================================\n",
      "Start time: 2026-02-18 12:34:18\n",
      "\n",
      "STEP 1: Loading sequencing reads\n",
      "--------------------------------------------------------------------------------\n",
      "Input file: data/mouse_SE_150bp.fq\n",
      "Expected: 10 million reads, 150bp each\n",
      "\n"
     ]
    },
    {
     "ename": "NameError",
     "evalue": "name 'read_fastq' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001B[31m---------------------------------------------------------------------------\u001B[39m",
      "\u001B[31mNameError\u001B[39m                                 Traceback (most recent call last)",
      "\u001B[36mCell\u001B[39m\u001B[36m \u001B[39m\u001B[32mIn[3]\u001B[39m\u001B[32m, line 327\u001B[39m\n\u001B[32m    324\u001B[39m \u001B[38;5;28mprint\u001B[39m(\u001B[33m\"\u001B[39m\u001B[33m=\u001B[39m\u001B[33m\"\u001B[39m*\u001B[32m80\u001B[39m + \u001B[33m\"\u001B[39m\u001B[38;5;130;01m\\n\u001B[39;00m\u001B[33m\"\u001B[39m)\n\u001B[32m    326\u001B[39m \u001B[38;5;66;03m# Run the assembly pipeline\u001B[39;00m\n\u001B[32m--> \u001B[39m\u001B[32m327\u001B[39m result = \u001B[43massemble_mouse_genome\u001B[49m\u001B[43m(\u001B[49m\u001B[43m)\u001B[49m\n\u001B[32m    329\u001B[39m \u001B[38;5;66;03m# Display sample contigs\u001B[39;00m\n\u001B[32m    330\u001B[39m \u001B[38;5;28mprint\u001B[39m(\u001B[33m\"\u001B[39m\u001B[33m=\u001B[39m\u001B[33m\"\u001B[39m*\u001B[32m80\u001B[39m)\n",
      "\u001B[36mCell\u001B[39m\u001B[36m \u001B[39m\u001B[32mIn[3]\u001B[39m\u001B[32m, line 173\u001B[39m, in \u001B[36massemble_mouse_genome\u001B[39m\u001B[34m(input_file, output_fasta, stats_file, k_mer_size, random_seed)\u001B[39m\n\u001B[32m    170\u001B[39m \u001B[38;5;28mprint\u001B[39m()\n\u001B[32m    172\u001B[39m start_time = time.time()\n\u001B[32m--> \u001B[39m\u001B[32m173\u001B[39m reads = \u001B[43mread_fastq\u001B[49m(input_file)\n\u001B[32m    174\u001B[39m timing[\u001B[33m'\u001B[39m\u001B[33mread_time\u001B[39m\u001B[33m'\u001B[39m] = time.time() - start_time\n\u001B[32m    176\u001B[39m num_reads = \u001B[38;5;28mlen\u001B[39m(reads)\n",
      "\u001B[31mNameError\u001B[39m: name 'read_fastq' is not defined"
     ]
    }
   ],
   "source": [
    "\"\"\"Driver program for mouse genome assembly using De Bruijn graphs.\n",
    "\n",
    "This script demonstrates how to use the completed DeBruijnGraph class to\n",
    "assemble 10 million perfect 150bp single-end reads from the mouse genome.\n",
    "\n",
    "Usage:\n",
    "    Run all cells in order in a Jupyter notebook.\n",
    "\n",
    "Expected input:\n",
    "    - File: mouse_SE_150bp.fq\n",
    "    - Format: FASTQ\n",
    "    - Reads: 10 million perfect 150bp single-end reads\n",
    "    - Source: Simulated from mouse genome using wgsim\n",
    "\n",
    "Output:\n",
    "    - mouse_assembly.fasta: Assembled contigs\n",
    "    - mouse_assembly_stats.txt: Detailed statistics\n",
    "\"\"\"\n",
    "\n",
    "\n",
    "print(\"\\n\" + \"=\"*80)\n",
    "print(\"MOUSE GENOME DE BRUIJN GRAPH ASSEMBLY\")\n",
    "print(\"Student Assignment Driver Program\")\n",
    "print(\"=\"*80 + \"\\n\")\n",
    "\n",
    "# Run the assembly pipeline\n",
    "result = assemble_mouse_genome()\n",
    "\n",
    "# Display sample contigs\n",
    "print(\"=\"*80)\n",
    "print(\"SAMPLE ASSEMBLED CONTIGS\")\n",
    "print(\"=\"*80 + \"\\n\")\n",
    "\n",
    "contigs = result['contigs']\n",
    "for i in range(min(5, len(contigs))):\n",
    "    contig = contigs[i]\n",
    "    preview = contig[:100] + \"...\" if len(contig) > 100 else contig\n",
    "    print(f\">contig_{i+1} length={len(contig)}\")\n",
    "    print(preview)\n",
    "    print()\n",
    "\n",
    "print(\"=\"*80)\n",
    "print(\"✓ ASSEMBLY COMPLETE\")\n",
    "print(\"=\"*80)\n",
    "print(f\"\\nResults saved to:\")\n",
    "print(f\"  • mouse_assembly.fasta - {result['stats']['num_contigs']:,} assembled contigs\")\n",
    "print(f\"  • mouse_assembly_stats.txt - Detailed assembly statistics\")\n",
    "print(f\"\\nKey metrics:\")\n",
    "print(f\"  • Total assembly: {result['stats']['total_length']:,} bp\")\n",
    "print(f\"  • N50: {result['stats']['n50']:,} bp\")\n",
    "print(f\"  • Longest contig: {result['stats']['longest_contig']:,} bp\")\n",
    "print(f\"  • Coverage: {result['coverage']:.1f}x\")\n",
    "print()"
   ]
  }
 ],
 "metadata": {
  "jupytext": {
   "formats": "ipynb,Rmd,qmd"
  },
  "kernelspec": {
   "display_name": "binf",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
