{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "5caec0a5-89c1-48b9-b10a-7ea6eda3fa8b",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pprint import pprint\n",
    "import numpy as np\n",
    "import random\n",
    "import motif_ops\n",
    "import seq_ops\n",
    "import seq_tools\n",
    "import seqlogo\n",
    "from scipy.special import softmax\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "977fefab-e4a7-4367-a1df-ae5478c3a6dc",
   "metadata": {},
   "outputs": [
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] No such file or directory: './data/only_peaks.bed'",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mFileNotFoundError\u001b[39m                         Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[2]\u001b[39m\u001b[32m, line 7\u001b[39m\n\u001b[32m      4\u001b[39m k = \u001b[32m8\u001b[39m\n\u001b[32m      5\u001b[39m ic_list=[]\n\u001b[32m----> \u001b[39m\u001b[32m7\u001b[39m seqs = \u001b[43mseq_tools\u001b[49m\u001b[43m.\u001b[49m\u001b[43mread_file\u001b[49m\u001b[43m(\u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43m./data/only_peaks.bed\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[32m     10\u001b[39m \u001b[38;5;66;03m#randomly choose motif start sites\u001b[39;00m\n\u001b[32m     11\u001b[39m motif_starts = seq_tools.build_motif_starts(seqs, k)\n",
      "\u001b[36mFile \u001b[39m\u001b[32m/mnt/c/Users/Eric Arnold/OneDrive/Documents/2026 Spring/BINF/EmpireOfDirt/project03/seq_tools.py:13\u001b[39m, in \u001b[36mread_file\u001b[39m\u001b[34m(filename)\u001b[39m\n\u001b[32m      7\u001b[39m \u001b[38;5;250m\u001b[39m\u001b[33;03m'''Takes file name and returns list of sequences found in file\u001b[39;00m\n\u001b[32m      8\u001b[39m \u001b[33;03mParams: filename (string) - name of file to be opened\u001b[39;00m\n\u001b[32m      9\u001b[39m \u001b[33;03mReturn: seqs (list) - list of strings of nucleotides\u001b[39;00m\n\u001b[32m     10\u001b[39m \u001b[33;03m'''\u001b[39;00m\n\u001b[32m     12\u001b[39m seqs = []\n\u001b[32m---> \u001b[39m\u001b[32m13\u001b[39m \u001b[38;5;28;01mwith\u001b[39;00m \u001b[38;5;28;43mopen\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43mfilename\u001b[49m\u001b[43m)\u001b[49m \u001b[38;5;28;01mas\u001b[39;00m file:\n\u001b[32m     14\u001b[39m     \u001b[38;5;28;01mfor\u001b[39;00m line \u001b[38;5;129;01min\u001b[39;00m file:\n\u001b[32m     15\u001b[39m         seqs.append(line.strip().strip(\u001b[33m\"\u001b[39m\u001b[33mN\u001b[39m\u001b[33m\"\u001b[39m))\n",
      "\u001b[31mFileNotFoundError\u001b[39m: [Errno 2] No such file or directory: './data/only_peaks.bed'"
     ]
    }
   ],
   "source": [
    "def GibbsMotifFinder(seqs, k, max_iterations=5000, threshold = 5e-10):\n",
    "\n",
    "    #initialize ic_list to track information content over iterations\n",
    "    ic_list = []\n",
    "\n",
    "    #randomly choose motif start sites\n",
    "    motif_starts = seq_tools.build_motif_starts(seqs, k)\n",
    "\n",
    "    #Intialize pfm, just the first time.\n",
    "    pfm = motif_ops.build_pfm(seq_tools.get_kmer_list(seqs, motif_starts, k), k)\n",
    "\n",
    "    # LOOP\n",
    "    i = 0 # count number of iterations\n",
    "    converged = False\n",
    "    while (converged != True) and (i <= max_iterations):\n",
    "        \n",
    "        #save a copy for next time to compare for convergence\n",
    "        pfm_old = pfm.copy()\n",
    "        \n",
    "        # select kmer to score\n",
    "        pick = random.randrange(len(motif_starts))\n",
    "        \n",
    "        # remove chosen kmer to score from pfm\n",
    "        seq_tools.change_pfm(seqs[pick][motif_starts[pick]:motif_starts[pick]+k], k, \"sub\", pfm)\n",
    "        \n",
    "        \n",
    "        #remove chosen kmer from list\n",
    "        removed_seq = seqs.pop(pick)\n",
    "        removed_index = motif_starts.pop(pick)\n",
    "        \n",
    "        #build pwm\n",
    "        pwm = motif_ops.build_pwm(pfm)\n",
    "        \n",
    "        #get reverse complement\n",
    "        rev_seq = seq_ops.reverse_complement(removed_seq)\n",
    "        \n",
    "        \n",
    "        #score kmers\n",
    "        kmer_scores = []\n",
    "        #score forward seq\n",
    "        for x in range(len(removed_seq) - k):\n",
    "            score = motif_ops.score_kmer(removed_seq[x:x+k], pwm)\n",
    "            kmer_scores.append(score)\n",
    "        \n",
    "        #score reverse seq\n",
    "        for x in range(len(removed_seq) - k):\n",
    "            score = motif_ops.score_kmer(rev_seq[x:x+k], pwm)\n",
    "            kmer_scores.append(score)\n",
    "        \n",
    "        \n",
    "        # choose \"best\" motif score using softmax calculation\n",
    "        prob = softmax(kmer_scores)\n",
    "        new_idx = np.random.choice(np.arange(len(kmer_scores)), p=prob)\n",
    "        \n",
    "        # use modulo to determine if chosen motif is on forward or reverse sequence\n",
    "        quotient, remainder = divmod(new_idx, len(removed_seq) - k - 1)\n",
    "        if quotient == 0: # forward sequence\n",
    "            new_motif = removed_seq[remainder:remainder+k]\n",
    "            seq_to_add_back = removed_seq\n",
    "        else: # reverse sequence\n",
    "            new_motif = rev_seq[remainder:remainder+k]\n",
    "            seq_to_add_back = rev_seq\n",
    "        \n",
    "        \n",
    "        # Insert the sequence and new index\n",
    "        seqs.append(seq_to_add_back)\n",
    "        motif_starts.append(remainder)\n",
    "        \n",
    "        # add new better motif to pfm\n",
    "        \n",
    "        seq_tools.change_pfm(new_motif, k, \"add\", pfm)\n",
    "        \n",
    "        # calculate information content and check for convergence\n",
    "        new_ic = motif_ops.pfm_ic(pfm)\n",
    "        old_ic = motif_ops.pfm_ic(pfm_old)\n",
    "        ic_diff = abs(new_ic-old_ic)\n",
    "        ic_list.append(new_ic)\n",
    "        \n",
    "        if ic_diff < threshold:\n",
    "            print(f\"\\rIteration {i}, we converged with IC of {ic_diff}\", end=\"\", flush=True)        \n",
    "            #converged = True\n",
    "        else: \n",
    "            print(f\"\\rIteration {i}, did not converge yet with IC of {ic_diff}\", end=\"\", flush=True)\n",
    "        i += 1\n",
    "    norm_pfm = pfm / pfm.sum(axis = 1, keepdims = True)\n",
    "    return norm_pfm, ic_list\n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "96f719af-1125-4726-af92-cb1781ace734",
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "the bed file is prepared separately using the following commands:\n",
    "\n",
    "# This file is generated by MACS version 3.0.3\n",
    "# Command line: callpeak -t ./data/SRR9090854.subsampled_5pct.bam -n output -f BAMPE --outdir ./macs3_output/ -g hs\n",
    "# ARGUMENTS LIST:\n",
    "# name = output\n",
    "# format = BAMPE\n",
    "# ChIP-seq file = ['./data/SRR9090854.subsampled_5pct.bam']\n",
    "# control file = None\n",
    "# effective genome size = 2.91e+09\n",
    "# band width = 300\n",
    "# model fold = [5, 50]\n",
    "# qvalue cutoff = 5.00e-02\n",
    "# The maximum gap between significant sites is assigned as the read length/tag size.\n",
    "# The minimum length of peaks is assigned as the predicted fragment length \"d\".\n",
    "# Larger dataset will be scaled towards smaller dataset.\n",
    "# Range for calculating regional lambda is: 10000 bps\n",
    "# Broad region calling is off\n",
    "# Paired-End mode is on\n",
    "\n",
    "# fragment size is determined as 9675 bps\n",
    "'''\n",
    "\n",
    "#load sequences from prepared bed file\n",
    "seqs = seq_tools.read_file(\"./data/only_peaks.bed\")\n",
    "\n",
    "#call the function\n",
    "norm_pfm, ic_list = GibbsMotifFinder(seqs, 8)\n",
    "\n",
    "pprint(norm_pfm.T)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "42a394c9-e584-4ae8-81e2-db3b89a1a48c",
   "metadata": {},
   "outputs": [],
   "source": [
    "seqlogo.seqlogo(seqlogo.CompletePm(pfm = norm_pfm.T), ic_scale=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c3d00763-585a-4ac1-8d7e-1974845447cd",
   "metadata": {},
   "outputs": [],
   "source": [
    "seqlogo.seqlogo(seqlogo.CompletePm(pfm = norm_pfm.T), ic_scale=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9021aa1f-aa0c-43e0-9af9-d55ad76fd15e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Visualize IC content over iterations\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "plt.plot(ic_list)\n",
    "plt.xlabel('Iterations')\n",
    "plt.ylabel('Score')\n",
    "plt.title('IC Content over time')\n",
    "plt.show()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "binf",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
