{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "bf8a9252-0371-46eb-983b-177fee981685",
   "metadata": {},
   "source": [
    "# Project 03: Gibbs Sampling (Random Algorithm)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "35409725-ccac-4013-bda5-ad8d03548608",
   "metadata": {},
   "outputs": [],
   "source": [
    "import random\n",
    "import numpy as np\n",
    "import bamnostic as bs\n",
    "import seqlogo as sl\n",
    "\n",
    "#import function for building sequence motif & idenfitying seqs matching to motif\n",
    "from data_readers import *\n",
    "from seq_ops import get_seq\n",
    "from motif_ops import *"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e6a1ad95-1019-42fc-839c-0a457decbedf",
   "metadata": {},
   "source": [
    "---\n",
    "## Implement Gibbs Sampler\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "51922616-8ff6-4d86-b3ef-ae714c03d1c4",
   "metadata": {},
   "source": [
    "Gibbs sampling is a MCMC approach to identify enrichments. Here we will implement a method to identify motifs from a set of regions. \n",
    "\n",
    "Important considerations:\n",
    "- We will need to score each sequence with a PWM using the `score_kmer()` or `score_sequence()` functions\n",
    "    - You will need to investigate into the help documenation and libraries to identify how best to use these functions. \n",
    "- These sites are often not strand-specific and so both scores on the negative as well as positive strand should be considered\n",
    "- To select a random sequence, use `random.randint()` or `numpy.random.randint()`\n",
    "- To select a new position $m$ (as defined below) use `random.choices()` or `numpy.random.choice()`  \n",
    "\n",
    "Assumptions: \n",
    "- We know $k$ as the length of expected motif\n",
    "- Each sequence contains the motif\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2601e5b1-6a3b-4f57-8eeb-e56f67cfda62",
   "metadata": {},
   "source": [
    "```\n",
    "GibbsMotifFinder(DNA, k-length)\n",
    "    random pick of k-length sequences from each line of DNA as Motifs\n",
    "    for j ← 1 to 10000 or Motifs stops changing\n",
    "        i ← Random(N) where N is number of DNA entries\n",
    "        PWM ← PWM constructed from all Motifs except for Motifi\n",
    "        Motifi ← select position m from PWM-scored k-mers in DNAi in probabilistic fashion from score distribution\n",
    "    return PFM\n",
    "```\n",
    "\n",
    "Probability of chosing position $m = \\frac{A_{m}}{\\sum_{l}A_{l}}$ for positions $l$ in DNAi\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1bcc0c5a-f32d-4a3a-9b3f-fc33edaf2e7c",
   "metadata": {},
   "source": [
    "**Note:** I have also added a function to `motif_ops.py` that will calculate the information content of your motifs. This is useful to observe the progression of your Gibbs sampler as well as a measure of convergence. You can use this function as `IC = pfm_ic(pfm)`. You should expect a slow increase of IC until it plateaus such as in the plot below from your lecture slides:\n",
    "\n",
    "<center><img src='figures/Gibbs_Sampling.png'/ width=600px></center>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5d218c76-f047-4e66-b1cc-3b1ffca510a0",
   "metadata": {},
   "source": [
    "---\n",
    "# Setting up the data\n",
    "\n",
    "This BAM file contains a subsampled set of aligned ChIP‑seq reads from a p53 immunoprecipitation experiment in human K562 cells treated with the anthracycline drug daunorubicin. The original SRA experiment SRX5865974 (run SRR9090854) reports 31.2 million Illumina NextSeq 500 ChIP‑seq read pairs from K562 wild‑type cells exposed to daunorubicin, using a p53 antibody to pull down p53‑bound chromatin fragments before library preparation and sequencing. [pmc.ncbi.nlm.nih](https://pmc.ncbi.nlm.nih.gov/articles/PMC4366240/)\n",
    "\n",
    "### What the data represent\n",
    "\n",
    "- Biological system: human K562 leukemia cells (Homo sapiens) treated with daunorubicin, a DNA‑damaging chemotherapeutic known to stabilize and activate p53. [pmc.ncbi.nlm.nih](https://pmc.ncbi.nlm.nih.gov/articles/PMC6561911/)\n",
    "- Assay: ChIP‑seq using an antibody against p53, so reads should be enriched around genomic regions where p53 is bound after drug treatment. [pmc.ncbi.nlm.nih](https://pmc.ncbi.nlm.nih.gov/articles/PMC4526040/)\n",
    "- Sequencing: Illumina NextSeq 500, with the raw run SRR9090854 corresponding to experiment SRX5865974. [github](https://github.com/ncbi/sra-tools/issues/213)\n",
    "- BAM file: `SRR9090854.subsampled_5pct.bam` is a coordinate‑sorted alignment file containing ~5% of the original mapped reads, typically created by random down‑sampling the full BAM to reduce file size and speed up exploratory analyses while preserving the overall distribution of p53 binding events. [ecseq](https://www.ecseq.com/support/ngs-snippets/how-to-extract-a-list-of-specific-read-IDs-from-a-BAM-file)\n",
    "\n",
    "### How `bamnostic` is used and what your code does\n",
    "\n",
    "The `bamnostic` package provides a pure‑Python interface to BAM files that mirrors the `pysam` API, including an `AlignmentFile` class whose iterator yields `AlignedSegment` objects representing individual aligned reads. In your code: [bamnostic.readthedocs](https://bamnostic.readthedocs.io/en/latest/bamnostic.html)\n",
    "\n",
    "```python\n",
    "bam_path = \"data/SRR9090854.subsampled_5pct.bam\"\n",
    "\n",
    "seqs = [read.seq for read in bs.AlignmentFile(bam_path)]\n",
    "```\n",
    "\n",
    "- `bs.AlignmentFile(bam_path)` opens the BAM file as an `AlignmentFile` object in binary read mode (default `'rb'`), reading the BAM header (reference contigs, read groups, etc.) and preparing a streaming interface to all aligned records. [github](https://github.com/betteridiot/bamnostic/blob/master/docs/source/quickstart.rst)\n",
    "- Iterating over `AlignmentFile` (`for read in bs.AlignmentFile(...)`) returns each aligned read as a `bamnostic.AlignedSegment` object, which exposes properties analogous to SAM fields such as query name, flags, reference name, position, mapping quality, CIGAR string, and the original sequencing **sequence**. [bamnostic.readthedocs](https://bamnostic.readthedocs.io/en/latest/bamnostic.html)\n",
    "- The `read.seq` attribute is the query (read) sequence string stored in the BAM, corresponding to the full read sequence (including any unaligned bases), as opposed to `query_alignment_sequence`, which would only contain the aligned portion. [bamnostic.readthedocs](https://bamnostic.readthedocs.io/en/latest/bamnostic.html)\n",
    "- The list comprehension `[read.seq for read in ...]` consumes the entire BAM stream and collects the nucleotide sequences from every subsampled ChIP‑seq read into a Python list `seqs`, which can then be used for downstream tasks such as motif discovery, k‑mer analysis, or quality checks on read content. [ucdavis-bioinformatics-training.github](https://ucdavis-bioinformatics-training.github.io/2022-Feb-Introduction-To-Python-For-Bioinformatics/python/python5)\n",
    "\n",
    "In summary, your dataset is a 5% random sample of p53‑ChIP‑seq alignments from daunorubicin‑treated K562 cells, and the `bamnostic` code opens the subsampled BAM and extracts the raw read sequences from each aligned fragment into memory as a list."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "82427797-e842-40f8-929d-9d768e1498b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "bam_path = \"data/SRR9090854.subsampled_5pct.bam\"\n",
    "\n",
    "seqs = [read.seq for read in bs.AlignmentFile(bam_path)]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f06d29d2-617a-4821-942c-b28975dc3e54",
   "metadata": {},
   "source": [
    "---\n",
    "# Project Start"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dccbe0a4-852c-4c15-a64b-893d00e7014f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def GibbsMotifFinder (seqs, k, seed=None):\n",
    "    '''\n",
    "    Function to find a pfm from a list of strings using a Gibbs sampler\n",
    "    \n",
    "    Args: \n",
    "        seqs (str list): a list of sequences, not necessarily in same lengths\n",
    "        k (int): the length of motif to find\n",
    "        seed (int, default=None): seed for np.random\n",
    "\n",
    "    Returns:\n",
    "        pfm (numpy array): dimensions are 4xlength\n",
    "    '''\n",
    "    # Use rng to make random samples/selections/numbers\n",
    "    # Example: randint = rng.integer(1, 10)\n",
    "    random.seed(seed)\n",
    "    rng = np.random.default_rng(seed)\n",
    "\n",
    "    pass"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eab1f791-6765-46fd-a695-2c6a78430a3c",
   "metadata": {},
   "source": [
    "---\n",
    "# Driver Program\n",
    "Don't change any of the code here. If you have completed the project by following the coding by contract, the following code should work."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0a730251-97cb-42df-aea9-eaccd0d03131",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Run the gibbs sampler:\n",
    "promoter_pfm = GibbsMotifFinder(seqs,10 )\n",
    "\n",
    "# Plot the final pfm that is generated: \n",
    "seqlogo.seqlogo(seqlogo.CompletePm(pfm = promoter_pfm.T))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (testing_bam_seq)",
   "language": "python",
   "name": "conda-env-testing_bam_seq-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.14.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
